presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/SqlInvokedFunctionNamespaceManagerConfig.java

```diff
@@ -14,16 +14,25 @@
 package com.facebook.presto.functionNamespace;
 
 import com.facebook.airlift.configuration.Config;
+import com.facebook.presto.spi.function.RoutineCharacteristics;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableSet;
 import io.airlift.units.Duration;
 import io.airlift.units.MinDuration;
 
+import java.util.Set;
+
+import static com.facebook.presto.spi.function.RoutineCharacteristics.Language.SQL;
 import static java.util.concurrent.TimeUnit.HOURS;
 import static java.util.concurrent.TimeUnit.MINUTES;
 
 public class SqlInvokedFunctionNamespaceManagerConfig
 {
+    private static final Splitter LANGUAGE_SPLITTER = Splitter.on(',').omitEmptyStrings().trimResults();
+
     private Duration functionCacheExpiration = new Duration(5, MINUTES);
     private Duration functionInstanceCacheExpiration = new Duration(8, HOURS);
+    private Set<RoutineCharacteristics.Language> supportedFunctionLanguages = ImmutableSet.of(SQL);
```

Comment:
This doesn't seems to be related to syntax support, and I suggest to move this to engine: CreateFunctionTask.

If I'm a community user and admins to some Presto clusters, I would prefer less configuration parameters to set and tune.

e.g.I want to use remote UDFs. I have set it up the UDF server, added the end points to the Presto engine config properties, and I already have function namespace manager setup because we already supported SQL functions. Why can't it "just work"? Can't the engine just figure it out the support languages?

If engine and remote UDF server supports SQL, JAVA, JS, it is meaningless for me as an admin to configure the support language "python" anyway. So all I do here is mechanically set this configuration to match with whatever is available on the UDF server, because Presto requires me to?

Reply:
The function namespace manager needs to configure which language it supports, together with how to invoke that type of functions. The execution of a function is in function namespace manager, so why the configuration should live in the engine? Just like if a specific function namespace manager manages a `catalog.schema`, so you would call that function namespace manager's `createFunction` when you create the function, you would call that function namespace manager's `executeFunction` when you invoke it. The function namespace manager also needs to configure the remote tier and protocol used to execute the function. I can't see why these logic should fit in the engine. You can have function namespace manager A which can support language L1, and L2, where for L1 it will execute the function in T1 and for L2 in T2, while having another function namespace manager supporting language L1 and L3 that runs them in T1' and T3. It's way easier to manage these in function namespace manager than in engine. Remote function is not something that will be able to work out of the box. You HAVE to setup a remote function server to run the functions. The function you introduce has tighter relationship to the UDF server and function metadata than to Presto engine.